src/utils.cpp

```diff
@@ -1077,15 +1077,49 @@ std::string str_join(const std::vector<std::string> &list,
   return str;
 }
 
-bool is_numeric(const std::string &s)
+std::optional<std::variant<int64_t, uint64_t>> get_int_from_str(
+    const std::string &s)
 {
-  std::size_t idx;
-  try {
-    std::stoll(s, &idx, 0);
-  } catch (...) {
-    return false;
+  if (s.size() == 0) {
+    return std::nullopt;
+  }
+
+  static const auto hex_regex = std::regex("0[xX][0-9a-fA-F]+");
+
+  if (std::regex_match(s, hex_regex)) {
+    // Treat all hex's as unsigned
+    std::size_t idx;
+    try {
+      uint64_t ret = std::stoull(s, &idx, 0);
+      if (idx == s.size()) {
+        return ret;
+      } else {
+        return std::nullopt;
+      }
+    } catch (...) {
+      // This shouldn't happen
+      return std::nullopt;
+    }
+  }
+
+  char *endptr;
+  errno = 0;
+
+  if (s.at(0) == '-') {
+    int64_t ret = strtol(s.c_str(), &endptr, 0);
+    if (endptr == s.c_str() || *endptr != '\0' || errno == ERANGE ||
+        errno == EINVAL) {
```

Comment:
Maybe defensively save a copy of `errno` incase `c_str()` clobbers it? It probably doens't, but who knows

Reply:
I'll just call `c_str()` before resetting errno and save the ptr.